File: 2015-02/hopscotch.py
import sys
sys.stdin = open("hopscotch.in")
# sys.stdout = open("hopscotch.out", 'w')
input = sys.stdin.readline

MOD = 10 ** 9 + 7

r, c, k = map(int, input().split())
board = [list(map(int, input().split())) for _ in range(r)]
# print(*board, sep='\n')

# jump from every square to the current one being processed
# TC: R**2 * C**2
# ok, there's probably a way to speed this up
dp = [[0 for _ in range(c)] for _ in range(r)]
dp[0][0] = 1
for i in range(1, r):
    for j in range(1, c):
        for x in range(i):
            for y in range(j):
                if board[i][j] != board[x][y]:
                    dp[i][j] += dp[x][y]
        dp[i][j] %= MOD
print(dp[-1][-1])

File: 2015-02/hopscotch.cpp
#include <iostream>
#include <vector>

using namespace std;

int r, c, k;
int x;
int main() {
    freopen("hopscotch.in", "r", stdin);
    freopen("hopscotch.out", "w", stdout);
    int MOD = 1000000007;
    cin >> r >> c >> k;
    vector<vector<int>> board(r, vector<int>(c));
    for (int i = 0; i < r; i++) {
        for (int j = 0; j < c; j++) {
            cin >> x;
            board[i][j] = x;
        }
    }
    // for (int i = 0; i < r; i++) {
    //     for (int j = 0; j < c; j++) {
    //         cout << board[i][j] << " ";
    //     }
    //     cout << endl;
    // }
    vector<vector<int>> dp(r, vector<int>(c));
    dp[0][0] = 1;
    for (int i = 0; i < r; i++) {
        for (int j = 0; j < c; j++) {
            for (int x = 0; x < i; x++) {
                for (int y = 0; y < j; y++) {
                    if (board[i][j] != board[x][y]) {
                        dp[i][j] += dp[x][y];
                        dp[i][j] %= MOD;
                    }
                }
            }
        }
    }
    cout << dp[r - 1][c - 1];
    return 0;
}
